lib/collection/src/collection_manager/optimizers/vacuum_optimizer.rs

```diff
@@ -64,26 +69,107 @@ impl VacuumOptimizer {
+    ///
+    /// The ratio is based on the number of deleted vectors in each named vector in which the
+    /// vacuum threshold is reached.
+    ///
+    /// Returns `None` if littered ratio did not reach vacuum thresholds for no named vectors.
+    fn littered_ratio_vectors(&self, segment: &LockedSegment) -> Option<f64> {
+        let segment_entry = segment.get();
+        let read_segment = segment_entry.read();
+
+        // Never optimize special segments
+        if read_segment.segment_type() == SegmentType::Special {
+            return None;
+        }
+
+        // Segment must have an index
+        let segment_config = read_segment.config();
+        if !segment_config.is_vector_indexed() {
+            return None;
+        }
+
+        // We can only work with original segments
+        let real_segment = match segment {
+            LockedSegment::Original(segment) => segment.read(),
+            LockedSegment::Proxy(_) => return None,
+        };
+        let total_vector_count = real_segment
+            .vector_data
+            .values()
+            .map(|vector_data| vector_data.vector_storage.borrow().total_vector_count())
+            .sum::<usize>();
+        if total_vector_count == 0 {
+            return None;
+        }
+
+        // In this segment, check the index of each named vector for a high deletion count.
+        // Count all deleted vectors of those that reach a threshold.
+        let deleted_vector_count = real_segment
+            .vector_data
+            .values()
+            .filter(|vector_data| vector_data.vector_index.borrow().is_index())
+            .filter_map(|vector_data| {
+                // We asume the storage and its index are created at the same time. We
+                // therefore know the number of deleted vectors then and now. Based on that we
+                // can determine what ratio of points from the index is deleted.
+                let vector_storage = vector_data.vector_storage.borrow();
+                let create_deleted_vec_count = vector_storage.create_deleted_vector_count();
+                let full_index_count =
+                    vector_storage.total_vector_count() - create_deleted_vec_count;
+                let deleted_index_count =
+                    vector_storage.deleted_vector_count() - create_deleted_vec_count;
+                let deleted_ratio = if full_index_count != 0 {
+                    deleted_index_count as f64 / full_index_count as f64
+                } else {
+                    0.0
+                };
+
+                let reached_minimum = deleted_index_count >= self.min_vectors_number;
+                let reached_ratio = deleted_ratio > self.deleted_threshold;
+                (reached_minimum && reached_ratio).then_some(deleted_index_count)
             })
-            .max_by_key(|(_, ratio)| OrderedFloat(*ratio))
-            .map(|(idx, _)| (idx, segments_read_guard.get(idx).unwrap().clone()))
+            .sum::<usize>();
```

Comment:
same, it is more relevant to use max

Reply:
Updated!

The idea was to select and sum all storages within a segment that reach the threshold, as the optimizer(s) work on a whole segment at once. But yes, using max makes sense so we pick the worst storage.